### ISSUE TITLE
missing privacy api

### ISSUE THREAD

**Issue by danmarsden** [duitkupg/moodle-paygw_duitku#2]
Moodle uses a privacy API for GDPR compliance to allow plugins to specify how they deal with user data. Your plugin stores user data in a number of tables which will need to be included in the privaci api classes.

As your plugin connects to an external integration, the privacy api is required for approval in the plugins db.

Sites that use continuous integration processes will not be able to use your plugin because Moodle runs unit tests which check to see if all extra plugins include the privacy class.

More information on the privacy class is here:
https://docs.moodle.org/dev/Privacy_API

**Comment by maikdi** [comment]
@danmarsden I'm new regarding this GDPR issue... Among all the functions that are in the Privacy API, what are the only functions I need to implement? Seeing how a famous payment gateway plugin (will not say the name, but it starts with S) only implemented the get_metadata() function. 

My plan is to implement these functions:
- get_metadata()
- get_contexts_for_userid()
- get_users_in_context()
- export_user_data()
- delete_data_for_all_users_in_context()
- delete_data_for_user()
- delete_data_for_users()

Is it ok to just implement the get_metadata() function then?

**Comment by maikdi** [comment]
TL;DR

Is it ok if I only implement the get_metadata() function (seeing how another payment gateway only implements that)?
Do I need to implement the export_payment_data() and delete_data_for_payment_sql() even though the plugin database does not store the paymentid? If I do need to implement it does that mean I should export all the user data that was used in the plugin or just my order id? (again taking references from other payment gateways)

**Comment by danmarsden** [comment]
@maikdi - the original version of the stripe plugin didn't store any user data within it's tables so it only had to implement the external_location_link stuff (although I do see a new intents table which contains a userid field so I've reported that in their github tracker. (The developer who looks after the stripe plugin will probably fix that up pretty quick.)

For Dukiktu, you will need to implement an external_location handler like the one you see in the stripe plugin, but you will also need to implement handlers for the user data in your paygw_duitku table so no, sorry, the get_metadata() function is not all you need here.

We are more strict on payment plugins/plugins that integrate with an external service and do require them to implement a correct privacy api class sorry!

thanks

**Comment by maikdi** [comment]
Thanks for the explanation :D

**Comment by maikdi** [comment]
@danmarsden 
My provider class implemented from these interfaces as they store user data and transactions
\core_privacy\local\metadata\provider,
\core_privacy\local\request\plugin\provider,
\core_privacy\local\request\core_userlist_provider

Will these suffice or are there other interfaces I must implement as well?

**Comment by danmarsden** [comment]
for the purposes of what is required for plugins db approval that sounds ok - as long as we can see add_external_location_link and the tables that contain user information look covered by your privacy api that should be ok.
